source/module_cusolver/test/CMakeLists.txt

```diff
@@ -0,0 +1,10 @@
+remove_definitions(-D__MPI)
```

Comment:
You are stripping MPI macro definitions from `mpi_matrix_test.cpp`, this might be confusing for the test relies on MPI.

Reply:
This file was based on the file source/module_cusolver/test/CMakeLists.txt and in fact I don't know well about the macro's role in unit test.
<img width="528" alt="image" src="https://user-images.githubusercontent.com/25193323/159231729-302d4b66-b036-4e00-82a4-6d2fdcada4d0.png">

Reply:
> This file was based on the file source/module_cusolver/test/CMakeLists.txt and in fact I don't know well about the macro's role in unit test. 

The MPI flag was set in the root CMakeLists, however enabling it will cause Unit Test build MPI-related code, which is suppported neither by GTest nor by our UTs. So we have to remove this flag.

Reply:
This file /source/module_base/test/tool_quit_test.cpp uses __MPI to activate parallel environment and I think GTest may be used for mpi tests to some entent while I'm not sure whether the `mpi_matrix_test.cpp` is coded in a proper way.
<img width="567" alt="image" src="https://user-images.githubusercontent.com/25193323/159403540-c260b76b-015b-485a-8f67-157102719b54.png">

Reply:
If we use MPI, some code relies on `src_parallel`, and this module includes global variables and constants, which forces us to link the whole project into one UT. This is obviously bad, and we are fixing it during our ongoing code refactor.
For much UTs, they don't need to concern about the parallel part, so disabling MPI flag works, and the UTs only tests the non-parallel part. For this one(`mpi_matrix_test`), do we need to test the split and the communication of data of MPI?

> This file /source/module_base/test/tool_quit_test.cpp uses __MPI to activate parallel environment and I think GTest may be used for mpi tests to some entent while I'm not sure whether the `mpi_matrix_test.cpp` is coded in a proper way. 

It seems there should be no MPI flag in this test, and those code would not take effects(see [CMakeLists](https://github.com/deepmodeling/abacus-develop/blob/develop/source/module_base/test/CMakeLists.txt#L1)). @hongriTianqi Could you please explain to us?

Reply:
> If we use MPI, some code relies on `src_parallel`, and this module includes global variables and constants, which forces us to link the whole project into one UT. This is obviously bad, and we are fixing it during our ongoing code refactor. For much UTs, they don't need to concern about the parallel part, so disabling MPI flag works, and the UTs only tests the non-parallel part. For this one(`mpi_matrix_test`), do we need to test the split and the communication of data of MPI?

Yes. What we want to test is whether the mpi communication part, i.e. RootGather and RootDivide we implemented, succeeded.

Reply:
> > If we use MPI, some code relies on `src_parallel`, and this module includes global variables and constants, which forces us to link the whole project into one UT. This is obviously bad, and we are fixing it during our ongoing code refactor. For much UTs, they don't need to concern about the parallel part, so disabling MPI flag works, and the UTs only tests the non-parallel part. For this one(`mpi_matrix_test`), do we need to test the split and the communication of data of MPI?
> 
> Yes. What we want to test is whether the mpi communication part, i.e. RootGather and RootDivide we implemented, succeeded.

I see. So, this MPI flag in CMakeLists doesn't matter, because in your test code those part are not controlled by `#ifdef __MPI`, am I right?
However, to launch this test executable, we need to specify launch comands as `mpirun`, rather than the default one. (See [here](https://github.com/deepmodeling/abacus-develop/blob/develop/CMakeLists.txt#L253)).

Reply:
> > > If we use MPI, some code relies on `src_parallel`, and this module includes global variables and constants, which forces us to link the whole project into one UT. This is obviously bad, and we are fixing it during our ongoing code refactor. For much UTs, they don't need to concern about the parallel part, so disabling MPI flag works, and the UTs only tests the non-parallel part. For this one(`mpi_matrix_test`), do we need to test the split and the communication of data of MPI?
> > 
> > 
> > Yes. What we want to test is whether the mpi communication part, i.e. RootGather and RootDivide we implemented, succeeded.
> 
> I see. So, this MPI flag in CMakeLists doesn't matter, because in your test code those part are not controlled by `#ifdef __MPI`, am I right? However, to launch this test executable, we need to specify launch comands as `mpirun`, rather than the default one. (See [here](https://github.com/deepmodeling/abacus-develop/blob/develop/CMakeLists.txt#L253), we don't have this feature yet).

Reply:
> > > If we use MPI, some code relies on `src_parallel`, and this module includes global variables and constants, which forces us to link the whole project into one UT. This is obviously bad, and we are fixing it during our ongoing code refactor. For much UTs, they don't need to concern about the parallel part, so disabling MPI flag works, and the UTs only tests the non-parallel part. For this one(`mpi_matrix_test`), do we need to test the split and the communication of data of MPI?
> > 
> > 
> > Yes. What we want to test is whether the mpi communication part, i.e. RootGather and RootDivide we implemented, succeeded.
> 
> I see. So, this MPI flag in CMakeLists doesn't matter, because in your test code those part are not controlled by `#ifdef __MPI`, am I right? However, to launch this test executable, we need to specify launch comands as `mpirun`, rather than the default one. (See [here](https://github.com/deepmodeling/abacus-develop/blob/develop/CMakeLists.txt#L253); we don't have this feature yet).

Yes, the `mpi_matrox_test.cpp` doesn't depend on the flag`__MPI` and it actually makes sense only when the test file  run in multiple processes. However, it nevertheless doesn't conflict with current launch command and can be reserved for future.